jmclient/jmclient/wallet_rpc.py

```diff
@@ -0,0 +1,674 @@
+                inner_dict = {}
+                inner_dict.update({"utxo": k})
+                for k2, v2 in v.items():
+                    inner_dict.update({k2: v2})
+                res.append(inner_dict)
+            return res
+
+        #route to list utxos
+        @app.route('/wallet/<string:walletname>/utxos',methods=['GET'])
+        def listutxos(self, request, walletname):
+            self.check_cookie(request)
+            if not self.wallet_service:
+                raise NoWalletFound()
+            # note: the output of `showutxos` is already a string for CLI;
+            # but we return json:
+            utxos = json.loads(wallet_showutxos(self.wallet_service, False))
+            utxos_response = self.get_listutxos_response(utxos)
+            return make_jmwalletd_response(request, utxos=utxos_response)
+
+        #route to start a coinjoin transaction
+        @app.route('/wallet/<string:walletname>/taker/coinjoin',methods=['POST'])
+        def docoinjoin(self, request, walletname):
+            self.check_cookie(request)
+            if not self.wallet_service:
+                raise NoWalletFound()
+            request_data = self.get_POST_body(request,["mixdepth", "amount",
+                                            "counterparties", "destination"])
+            if not request_data:
+                raise InvalidRequestFormat()
+            #see file scripts/sample-schedule-for-testnet for schedule format
+            waittime = 0
+            rounding=16
+            completion_flag=0
+            # A schedule is a list of lists, here we have only one item
+            schedule = [[int(request_data["mixdepth"]), int(request_data["amount"]),
+                         int(request_data["counterparties"]),
+                         request_data["destination"], waittime, rounding,
+                         completion_flag]]
+            #instantiate a taker
+            #keeping order_chooser as default for now
+            #max_cj_fee is to be set based on config values
+            # TODO: must set this from config (POLICY section, `max_cj_fee_rel/abs`.
+            max_cj_fee=(1,float('inf'))
+            self.taker = Taker(self.wallet_service, schedule,
+                               max_cj_fee = max_cj_fee, callbacks=(
+                        self.filter_orders_callback, None,  self.taker_finished))
+            clientfactory = JMClientProtocolFactory(self.taker)
+            nodaemon = jm_single().config.getint("DAEMON", "no_daemon")
+            daemon = True if nodaemon == 1 else False
+            dhost = jm_single().config.get("DAEMON", "daemon_host")
+            dport = jm_single().config.getint("DAEMON", "daemon_port")
+
+            if jm_single().config.get("BLOCKCHAIN", "network") == "regtest":
+                startLogging(sys.stdout)
+            start_reactor(dhost, dport, clientfactory, daemon=daemon, rs=False)
+            return make_jmwalletd_response(request, coinjoin_started=True)
```

Comment:
Tried this method out and I was able to do coinjoin through RPC, with the `session` method signaling it (i.e., `coinjoin_in_process` is True`), though a couple of things:
* if we get an error while processing the coinjoin (e.g., `ERROR not enough liquidity in the orderbook`), then the entire RPC server shuts down and it needs to be restarted, same if the coinjoin completes successfully. i.e., after seeing `All transactions completed correctly` (or an error), JM disconnects from IRC and shuts down.
* `coinjoin_in_process` remains set to True until JM server shuts down, as per above.  I think this flag should become True right after the coordination part is done (i.e., we have broadcasted the transaction), without waiting for it to confirm.

Reply:
> if we get an error while processing the coinjoin (e.g., ERROR not enough liquidity in the orderbook), then the entire RPC server shuts down and it needs to be restarted, same if the coinjoin completes successfully. i.e., after seeing All transactions completed correctly (or an error), JM disconnects from IRC and shuts down.

This should now be addressed after https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/996/commits/26ddc758939bf8ff2be84e3675b693e88e5ce9b8 - see how the taker_finished callback now does not stop the reactor, but note also as per comments, that we are keeping it restricted to single joins for now. Just don't want to bite off more than we can chew (same goes for adding a bunch more API calls - can be done later).

There's also a primitive test of the /coinjoin route, though it doesn't do the whole thing.

> coinjoin_in_process remains set to True until JM server shuts down, as per above.

This was also changed using the `taker_finished` callback (called the `on_finished_callback` inside `Taker`).

Note the syntax was changed ("amount_sats"); see the commit comment.